presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +80,261 @@
+                    });
     }
 
     @Override
     public void close()
     {
-        queryExecutor.close();
+        if (closed.compareAndSet(false, true)) {
+            queryExecutorCleanupService.shutdown();
+            queryExecutorCache.invalidateAll();
+            jettyIoPool.close();
+        }
+    }
+
+    private class QueryExecutorHolder
+    {
+        private QueryExecutor queryExecutor;
+
+        private Map<Object, Object> clientProperties;
+
+        private AtomicInteger refcount = new AtomicInteger(0);
+
+        private QueryExecutorHolder(QueryExecutor queryExecutor, Map<Object, Object> clientProperties)
+        {
+            this.queryExecutor = requireNonNull(queryExecutor, "wrapped is null");
+            this.clientProperties = requireNonNull(clientProperties, "clientProperties is null");
+        }
+
+        private void closeQueryExecutor()
+        {
+            queryExecutor.close();
+        }
+
+        public void release()
+        {
+            Runnable deferredClose = () -> {
+                cacheLock.writeLock().lock();
+                try {
+                    if (refcount.decrementAndGet() == 0) {
+                        queryExecutorCache.invalidate(clientProperties);
+                    }
+                }
+                finally {
+                    cacheLock.writeLock().unlock();
+                }
+            };
+
+            /*
+             * Defer cleanup into the future to avoid constantly creating and
+             * destroying QE's in the case where a single thread is repeatedly
+             * calling connect() and close() for the same server.
+             */
+            queryExecutorCleanupService.schedule(deferredClose, 5, TimeUnit.MINUTES);
+        }
+
+        private QueryExecutor acquire()
+        {
+            refcount.incrementAndGet();
+            return queryExecutor;
+        }
+    }
+
+    private static Map<Object, Object> filterClientProperties(Properties connectionProperties)
+    {
+        /*
+         * TODO: Once there are client-specifc properties, return the subset of
+         * connectionProperties that is client-specific.
+         */
+        return ImmutableMap.copyOf(connectionProperties);
     }
 
     @Override
-    public Connection connect(String url, Properties info)
+    public Connection connect(String url, Properties connectionProperties)
             throws SQLException
     {
+        if (closed.get()) {
+            throw new SQLException("Already closed");
+        }
+
         if (!acceptsURL(url)) {
             return null;
         }
 
-        String user = info.getProperty(USER_PROPERTY);
+        String user = connectionProperties.getProperty(USER_PROPERTY);
         if (isNullOrEmpty(user)) {
             throw new SQLException(format("Username property (%s) must be set", USER_PROPERTY));
         }
 
-        return new PrestoConnection(new PrestoDriverUri(url), user, queryExecutor);
+        cacheLock.readLock().lock();
```

Comment:
It's a read-only operation in the sense that the existence of the QE in the cache is not altered. Removal takes the write lock because that is not true.